// @nestjs/platform-socket.io/adapters/io-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const shared_utils_1 = require("@nestjs/common/utils/shared.utils");
const websockets_1 = require("@nestjs/websockets");
const constants_1 = require("@nestjs/websockets/constants");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const io = require("socket.io");
class IoAdapter extends websockets_1.AbstractWsAdapter {
    create(port, options) {
        if (!options) {
            return this.createIOServer(port);
        }
        const { namespace, server } = options, opt = tslib_1.__rest(options, ["namespace", "server"]);
        return server && shared_utils_1.isFunction(server.of)
            ? server.of(namespace)
            : namespace
                ? this.createIOServer(port, opt).of(namespace)
                : this.createIOServer(port, opt);
    }
    createIOServer(port, options) {
        if (this.httpServer && port === 0) {
            return io(this.httpServer, options);
        }
        return io(port, options);
    }
    bindMessageHandlers(client, handlers, transform) {
        const disconnect$ = rxjs_1.fromEvent(client, constants_1.DISCONNECT_EVENT).pipe(operators_1.share(), operators_1.first());
        handlers.forEach(({ message, callback }) => {
            const source$ = rxjs_1.fromEvent(client, message).pipe(operators_1.mergeMap((payload) => {
                const { data, ack } = this.mapPayload(payload);
                return transform(callback(data, ack)).pipe(operators_1.filter((response) => !shared_utils_1.isNil(response)), operators_1.map((response) => [response, ack]));
            }), operators_1.takeUntil(disconnect$));
            source$.subscribe(([response, ack]) => {
                if (response.event) {
                    return client.emit(response.event, response.data);
                }
                shared_utils_1.isFunction(ack) && ack(response);
            });
        });
    }
    mapPayload(payload) {
        if (!Array.isArray(payload)) {
            return { data: payload };
        }
        const lastElement = payload[payload.length - 1];
        const isAck = shared_utils_1.isFunction(lastElement);
        if (isAck) {
            const size = payload.length - 1;
            return {
                data: size === 1 ? payload[0] : payload.slice(0, size),
                ack: lastElement,
            };
        }
        return { data: payload };
    }
}
exports.IoAdapter = IoAdapter;
